src/View/Shortcodes/EmbedShortcodeProvider.php

```diff
@@ -84,32 +86,27 @@ public static function handle_shortcode($arguments, $content, $parser, $shortcod
             $serviceArguments['min_image_height'] = $arguments['height'];
         }
 
-        /** @var EmbedResource $embed */
-        $embed = Injector::inst()->create(Embeddable::class, $serviceURL);
-        if (!empty($serviceArguments)) {
-            $embed->setOptions(array_merge($serviceArguments, (array) $embed->getOptions()));
+        /** @var Embeddable $embeddable */
+        // Allow mocking of Embeddable
+        if (($extra['Embeddable'] ?? null) instanceof Embeddable) {
+            $embeddable = $extra['Embeddable'];
```

Comment:
Yes Embeddable object with an Embeddable key

I'm not sure what you mean by `use Injector to override it with your mock`.  We don't want (can't?) mock EmbedShortcodeProvider because EmbedShortCodeProviderTest calls the static `EmbedShortcodeProvider::handle_shortcode()`, so I don't think mocks work here?

Reply:
I spend some time trying to get this to work, though what it comes down to is that EmbedContainer isn't a service class / singleton

We *could* refactor everything so that instead of `Injector::inst()->create(Embeddable::class, $url);` we use `$embeddable = Injector::inst()->get(Embeddable::class);` and then `$embeddable->setUrl($url);`.  And then all the other things that needs to be refactored.  And of course we need to add `public function setUrl(string $url)` to Embeddable, except we can't modify the interface, so that goes on to EmbedContainer only, which is messy since we're calling `Injector::inst()->get(Embeddable::class)`

We're not making things worse here, it was originally like this after all.  This is not work fussing with further IMO

Reply:
OK, have done this using the `Crawler` and the mockable/swappble bit